src/Engine/Segments/SegmentEvaluator.php

```diff
@@ -0,0 +1,187 @@
+     */
+    public static function evaluateIdentityInSegment(
+        IdentityModel $identity,
+        SegmentModel $segment,
+        array $overrideTraits = null
+    ): bool {
+        $rulesCount = count($segment->getRules());
+
+        if (empty($rulesCount)) {
+            return false;
+        }
+
+        $traitSegmentRules = [];
+        foreach ($segment->getRules() as $rule) {
+            $segmentRule = self::traitsMatchSegmentRule(
+                $overrideTraits ?? $identity->getIdentityTraits()->getArrayCopy(),
+                $rule,
+                $segment->getId(),
+                $identity->compositeKey()
+            );
+
+            $traitSegmentRules[] = $segmentRule;
+
+            if (!$segmentRule) {
+                break;
+            }
+        }
+
+        return self::all($traitSegmentRules);
+    }
+
+    /**
+     * traits match segment rule.
+     * @param array $identityTraits
+     * @param SegmentRuleModel $rule
+     * @param mixed $segmentId
+     * @param mixed $identityId
+     * @return bool
+     */
+    private static function traitsMatchSegmentRule(
+        array $identityTraits,
+        SegmentRuleModel $rule,
+        $segmentId,
+        $identityId
+    ): bool {
+        $traitConditions = [true];
+        $conditions = $rule->getConditions()->getArrayCopy();
+
+        if (count($conditions) > 0) {
+            $traitConditions = array_map(fn ($condition) => self::traitsMatchSegmentCondition(
+                $identityTraits,
+                $condition,
+                $segmentId,
+                $identityId
+            ), $conditions);
+        }
+
+        $matchesCondition = $rule->matchingFunction()($traitConditions);
+
+        $segmentRules = array_map(
+            fn ($ruleNew) => self::traitsMatchSegmentRule($identityTraits, $ruleNew, $segmentId, $identityId),
+            $rule->getRules()->getArrayCopy()
+        );
+
+        return $matchesCondition && (self::all($segmentRules));
+    }
+
+    /**
+     * traits that match the segment condition.
+     * @param array $identityTraits
+     * @param SegmentConditionModel $condition
+     * @param mixed $segmentId
+     * @param mixed $identityId
+     * @return bool
+     */
+    private static function traitsMatchSegmentCondition(
+        array $identityTraits,
+        SegmentConditionModel $condition,
+        $segmentId,
+        $identityId
+    ): bool {
+        if ($condition->getOperator() === SegmentConditions::PERCENTAGE_SPLIT) {
+            $value = floatval($condition->getValue());
+            return (self::getHashObject()->getHashedPercentageForObjectIds(
+                [$segmentId, $identityId]
+            ) <= $value);
+        }
+
+        $evaluations = array_filter(
+            $identityTraits,
+            fn (TraitModel $it) =>
+            $it->getTraitKey() === $condition->getProperty()
+        );
+
+        $trait = count($evaluations) > 0 ? array_shift($evaluations) : false;
```

Comment:
Again, I think this could be named better - it seems like this is evaluating to a bool but it's called `trait` ? I don't quite follow.

Reply:
This is extracting a matching trait equivalent to to the condition. Renamed it to $matchingTrait